cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1354,11 +1354,16 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
         throw new ApplicationNotFoundException(identifier.getApplication());
       } else if (!store.programExists(identifier)) {
         throw new ProgramNotFoundException(identifier);
-      } else if (runId == null) {
-        throw new BadRequestException("Program not running");
-      } else {
-        throw new NotFoundException(new Id.Run(identifier, runId));
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord.getProperties().containsKey("workflowrunid")
+          && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+          throw new BadRequestException("MapReduce program run started by Workflow cannot be stopped. " +
+                                          "Please stop the Workflow.");
+        }
       }
+      throw new NotFoundException(new Id.Run(identifier, runId));
```

Comment:
Also, you're not properly responded with BadRequestException("Program not running") like before. Isn't that still a possibility?